### ISSUE TITLE
QA的前复权算法有问题吗？

### ISSUE THREAD

**Issue by 5267** [yutiansut/QUANTAXIS#1333]
# 问题反馈

```
股票行情日线前复权和通达信前复权数据有差异
import time
import QUANTAXIS as QA
local_day = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(time.time())).split(' ')[0]
stock_ohlc = QA.QA_fetch_stock_day_adv(code = ['300033'], start = '1990-01-01', end = local_day)
stock_ohlc = stock_ohlc.to_qfq()
stock_ohlc_parts = stock_ohlc[(stock_ohlc.date>'2019-03-20') & (stock_ohlc.date<'2019-04-05')]()
print(stock_ohlc_parts )
```

今天，同事在使用前复权数据的时候，发现QA提供的to_qfq方法算出来的前复权数据和通达信的数据不一样，以同花顺为例，同花顺最近一次的除权除息日期是在20190329，我们看到前复权后，20190328之前的高开低收数据和同花顺的就有出入了。
![image](https://user-images.githubusercontent.com/19160836/67470083-c94c2180-f67f-11e9-81a7-fa23195b631d.png)

不知道，有没有其他小伙伴发现这个问题，特此推个issue，求验证。

**Comment by yutiansut** [comment]
本来就不一样呀...

通达信的算法有问题 详见群文件的 [wind复权算法.pdf]

QA用的是wind的算法

**Comment by 5267** [comment]
啊，通达信的算法有问题？ 全国人民用的通达信前复权数据居然有bug，次哦。我去瞅瞅wind的复权算法

**Comment by yutiansut** [comment]
@5267 反正按通达信的做法 1991年股票价格是负数 买股票送钱...  你看下wind的算法吧  群文件搜一下

**Comment by 5267** [comment]
嗯，正在享用 wind复权算法

**Comment by yutiansut** [comment]
另外 写法太神奇了stock_ohlc_parts = stock_ohlc[(stock_ohlc.date>'2019-03-20') & (stock_ohlc.date<'2019-04-05')]()
print(stock_ohlc_parts )

这是啥???

**Comment by yutiansut** [comment]
@5267 请这么写....

```python
stock_ohlc_parts = stock_ohlc.select_time('2019-03-20','2019-04-05')

```

**Comment by 5267** [comment]
哈哈，收到收到

**Comment by 5267** [comment]
问题还是有点多，前复权不一样的话，很多指标的值就和软件对不上了，所以，后续我们估计会再写一个通达信版本的复权公式，哭！

**Comment by yutiansut** [comment]
跟wind就能对上啊....

不过你要写个通达信的复权算法倒是也简单  多支持个复权选择 emm 好像也不错




------------------ Original ------------------

**Comment by 5267** [comment]
嗯，毕竟通达信的用户群体大啊，哈哈，目前我看各大主流行情商的复权算法都是和通达信一致的，嗷哦

**Comment by abc100m** [comment]
我也发现了, 这个库的前复权有BUG,  抽了一个K线看,跟sina算的是一样的,但是跟通达信的对不上. 仔细上看是这个库的不对

**Comment by abc100m** [comment]
以某日收盘价10.04, 14.5为例, 进行分红0.065除权
(10.04 - 0.065)     = 9.975
(14.5 - 0.065)      = 14.435
(10.04 - 0.065) / 10.04 * 14.5 = 14.406  

这个算法理论上是对的,但为啥结果就有BUG, 是不满足传递率还是精度丢失了数据?

--------------------
重新看了一下,貌似算法不一样

**Comment by yutiansut** [comment]
> 我也发现了, 这个库的前复权有BUG, 抽了一个K线看,跟sina算的是一样的,但是跟通达信的对不上. 仔细上看是这个库的不对

注意你自己的结束时间 群里有wind复权算法 这问题自己看pdf吧

**Comment by yutiansut** [comment]
> 我也发现了, 这个库的前复权有BUG, 抽了一个K线看,跟sina算的是一样的,但是跟通达信的对不上. 仔细上看是这个库的不对

tdx的复权算法自然的会导致2000年附近的价格是负数 我们只以wind为准  另外前复权取决于你最后一个bar的价格 qa使用的是动态复权 避免你在做2014-2017的回测用的是2020年收盘价算出来的复权价 这种就是典型的带未来信息的假回测

**Comment by abc100m** [comment]
请问一下月K线的前复权, 要怎么算呢? 我发现并不能简单的使用日K线前复权的代码去做。

TDX取到的原始数据（这是000002万科2019年8月的月K线）:
open,close,high,low
#-----------------------------
28.5,25.8,28.75,25.65,8659552.0,23369269248.0,2019-08-30,000002,1567094400.0

2019-08-15这一天分红: 10派10.451

TDX软件的前复权数据是:
open,close,high,low
#-----------------------------
27.45,25.80,27.77,25.36,....

这个涉及到月中分红，由取到的月K线原始数据是无法算出来的？ 
如果能的话，QUANTAXIS计算月K线的前复权代码在哪呢?
同样周K线的前复权，如果在某周的中间那天分红......

-------------
后记：QUANTAXIS 有resample的相关代码，#710，学习去~~~

**Comment by yutiansut** [comment]
是的  目前qa是使用 日线的DataStruct .to_qfq()  +  .week  实际上是在日线级别复权再resample的
